src/tasks/aiExtractor.ts

```diff
@@ -0,0 +1,153 @@
+import { TaskConfig } from 'payload'
+import { createGoogleGenerativeAI } from '@ai-sdk/google'
+import { generateObject } from 'ai'
+import { z } from 'zod'
+import { randomUUID } from 'node:crypto'
+
+export const AIExtractor: TaskConfig<'aiExtractor'> = {
+  slug: 'aiExtractor',
+  label: 'AI Extractor',
+  handler: async ({ req }) => {
+    const { payload } = req
+    const { logger } = payload
+    logger.info('aiExtractor:: Starting AI Extraction')
+
+    try {
+      const {
+        ai: { model: defaultModel, apiKey },
+      } = await payload.findGlobal({
+        slug: 'settings',
+      })
+
+      const google = createGoogleGenerativeAI({
+        apiKey: apiKey,
+      })
+
+      const model = google(defaultModel)
+
+      const systemPrompt = `
+    You are a helpful assistant that analyzes documents to extract campaign promises and provide structured summaries.
+    Your task is to identify and categorize campaign promises, providing direct quotations as sources.
+    `
+
+      const prompt = `
+    Analyze the following document and extract campaign promises with the following structure:
+
+    **Requirements:**
+    1. **Title**: Infer an appropriate title from the document content.
+    2. **Promises**: For each promise identified, provide:
+       - **Category**: The thematic category of the promise (e.g., "Economy," "Healthcare," "Infrastructure," "Education," "Social Policy," etc.)
+       - **Summary**: A concise summarization of the promise in your own words
+       - **Source**: Direct quotations from the text that support this promise. Include multiple quotes if the promise is mentioned in different parts of the document.
+
+    **Guidelines:**
+    - Only include specific, actionable pledges about future actions
+    - Exclude political commentary, statements of fact, or general values
+    - Source field must contain exact quotes from the document text
+    - Each promise should have clear supporting evidence in the source quotes
+    - Group similar promises under appropriate categories
+    - Reply in the language the document is written in
+
+    **Document:**
+    
+    `
+
+      const { docs: allDocs } = await payload.find({
+        collection: 'documents',
+        limit: -1,
+      })
+
+      const documents = allDocs.filter((doc) => doc.aiExtraction?.length === 0)
+
+      if (documents.length === 0) {
+        logger.info('aiExtractor:: No documents to Extract Promises')
+        return { output: { docs: [] } }
+      }
+
+      logger.info(`aiExtractor:: Extracting ${documents.length} documents`)
+
+      const processedDocs = []
+
+      for (const document of documents) {
```

Comment:
Why not use `documents.forEach((document) => { /*Logic block*/} )`

Reply:
@koechkevin, I used `for...of` because we’re running async functions inside the loop and need them to execute sequentially. `forEach` doesn’t wait for async operations to finish, so it’s unsuitable in this case. More context: https://stackoverflow.com/a/37576787